File: Backtracking/N_Queen/CS_N_Queen/N_Queen.cs
using System;

namespace CS_N_Queen
{
    class N_Queen
    {
        static int n, ans;
        static bool[] height, downTopDiagonal, topDownDiagonal;

        static void NQueen(int i)
        {
            if (i == n)
            {
                ans++;
                return;
            }

            for (int j = 0; j < n; j++)
            {
                if (height[j] || downTopDiagonal[i + j] || topDownDiagonal[i - j + n - 1])
			        continue;

                height[j] = true;
                downTopDiagonal[i + j] = true;
                topDownDiagonal[i - j + n - 1] = true;
		        NQueen(i + 1);
		        height[j] = false;
                downTopDiagonal[i + j] = false;
                topDownDiagonal[i - j + n - 1] = false;
	
            }
        }

        static void Main(string[] args)
        {
            n = Convert.ToInt32(Console.ReadLine());

            height = new bool[n];
            downTopDiagonal = new bool[2 * n - 1];
            topDownDiagonal = new bool[2 * n - 1];

            NQueen(0);

            Console.WriteLine(ans);
        }
    }
}

File: Backtracking/N_Queen/Python_N_Queen/N_Queen.py
import sys


n, ans = int(sys.stdin.readline()), 0
height = [False] * n
down_top_diagonal = [False] * (2 * n - 1)
top_down_diagonal = [False] * (2 * n - 1)


def n_queen(i):
    global ans

    if i == n:
        ans += 1
        return

    for j in range(n):
        if height[j] or down_top_diagonal[i+j] or top_down_diagonal[i-j+n-1]:
            continue

        height[j] = down_top_diagonal[i+j] = top_down_diagonal[i-j+n-1] = True
        n_queen(i + 1)
        height[j] = down_top_diagonal[i+j] = top_down_diagonal[i-j+n-1] = False


n_queen(0)

print(ans)
